tests/utils/test_shell_util.py

```diff
@@ -127,92 +130,237 @@ def test_it_should_log_unexpected_errors_as_errors(self):
         self.assertEqual(mock_logger.warn.call_count, 0)
 
 
-class RunCommandTestCase(AgentTestCase):
+# R0904: Too many public methods (24/20)  -- disabled: each method is a unit test
+class RunCommandTestCase(AgentTestCase):  # pylint: disable=R0904
+    """
+    Tests for shellutil.run_command/run_pipe
+    """
+    def __create_tee_script(self, return_code=0):
+        """
+        Creates a Python script that tees its stdin to stdout and stderr
+        """
+        tee_script = os.path.join(self.tmp_dir, "tee.py")
+
+        AgentTestCase.create_script(tee_script, """
+import sys
+
+for line in sys.stdin:
+    sys.stdout.write(line)
+    sys.stderr.write(line)
+exit({0})
+    """.format(return_code))
+
+        return tee_script
+
     def test_run_command_should_execute_the_command(self):
         command = ["echo", "-n", "A TEST STRING"]
         ret = shellutil.run_command(command)
         self.assertEqual(ret, "A TEST STRING")
 
-    def test_run_command_should_raise_an_exception_when_the_command_fails(self):
-        command = ["ls", "-d", "/etc", "nonexistent_file"]
+    def test_run_pipe_should_execute_a_pipe_with_two_commands(self):
+        # Output the same string 3 times and then remove duplicates
+        test_string = "A TEST STRING\n"
+        pipe = [["echo", "-n", "-e", test_string * 3], ["uniq"]]
+
+        output = shellutil.run_pipe(pipe)
+
+        self.assertEqual(output, test_string)
 
+    def test_run_pipe_should_execute_a_pipe_with_more_than_two_commands(self):
+        #
+        # The test pipe splits the output of "ls" in lines and count the number of lines.
+        #
+        # Sample output of "ls":
+        #     drwxrwxr-x 13 nam nam 4096 Nov 13 16:54 .
+        #
+        pipe = [["ls", "-ld", "."], ["sed", "s/ /\\n/g"], ["wc", "-l"]]
+
+        output = shellutil.run_pipe(pipe)
+
+        self.assertEqual(int(output), 9)
```

Comment:
Will this always be 9? Like we might add some file later which will break this test right? 
It might be a better idea to create a temp directory and add some random pre-determined files there and ls in that to ensure any changes to this directory wont affect the test

Reply:
I am using a pre-determined file (directory, actually): "." so this wouldn't change

Reply:
Ohh got it! I had to run the command to see the output lol.

Reply:
Oh, I see, I think know where you are coming from. I am using the -d option, so this does not list the contents of ".", but instead it lists the info for "." itself. This always produces only 1 line of output so this won't change.

Reply:
I had the same thought process and I had to run the command to figure out what it was doing :smile: